## Importing Libraries

In [1]:
# import libraries
import pandas as pd
import numpy as np
import os

In [2]:
# create a path variable
path = r'C:\Users\Sreelakshmi\Desktop\Instacart Basket Analysis'

In [3]:
#import the dataset ords_prods_merge to the dataframe
df_ords_prods_merge = pd.read_pickle(os.path.join(path,'Data', 'Prepared Data', 'ords_prods_merge.pkl'))

In [4]:
df_ords_prods_merge.shape

(32404859, 14)

In [5]:
# create a subset of this dataframe which contains first 1000000 rows
df = df_ords_prods_merge[:1000000]

In [6]:
df.shape

(1000000, 14)

In [7]:
df.head()

order_id  user_id  order_number  order_days_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     7.0         196                  1          0   
1                     7.0       14084                  2          0   
2                     7.0       12427                  3          0   
3                     7.0       26088                  4          0   
4                     7.0       26405                  5          0   

                              product_name  aisle_id  department_id  prices  \
0                                     Soda        77              7     9.0   
1  Organic Unsweetened Vanilla Almond Milk        91             16    12.5   
2                      Original Beef Jerky        23             19     4.4   
3               Aged White Cheddar Popcorn        23             19     4.7   
4         XL Pick-A-Size Paper Towel Rolls        54             17     1.0   

  _merge  
0   both  
1   both  
2   both  
3   both  
4   both

## Defining a user-defined function with if-else conditions

In [8]:
# function definition
def price_label(row):
    if row['prices'] <= 5:
        return 'low_range_product'
    elif (row['prices'] > 5 and row['prices'] <= 15):
        return 'mid_range_product'
    elif row['prices'] > 15:
        return 'high_range_product'
    else: return 'Not enough data'

In [9]:
# calling the function b creating a new column in the df dataframe
df['price_range'] = df.apply(price_label, axis=1)

C:\Users\Sreelakshmi\AppData\Local\Temp\ipykernel_13624\2494791123.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['price_range'] = df.apply(price_label, axis=1)


In [10]:
df['price_range'].value_counts(dropna = False)

price_range
mid_range_product     673183
low_range_product     314392
high_range_product     12425
Name: count, dtype: int64

### To find out maximum value in a particular column

In [11]:
# use max() function
df['prices'].max()

99999.0

### If-Statements with the loc() function

In [44]:
df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'

In [13]:
df.loc[(df['prices'] <= 15) & (df['prices'] > 5), 'price_range_loc'] = 'Mid-range product'

In [14]:
df.loc[df['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [15]:
df['price_range_loc'].value_counts()

price_range_loc
Mid-range product     673183
Low-range product     314392
High-range product     12425
Name: count, dtype: int64

#### Using loc() function in the entire dataset

In [16]:
df_ords_prods_merge.loc[df_ords_prods_merge['prices'] > 15, 'price_range_loc'] = 'High-range product'

In [17]:
df_ords_prods_merge.loc[(df_ords_prods_merge['prices'] <= 15) & (df['prices'] > 5), 'price_range_loc'] = 'Mid-range product'

In [18]:
df_ords_prods_merge.loc[df_ords_prods_merge['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [19]:
df_ords_prods_merge['price_range_loc'].value_counts()

price_range_loc
Low-range product     10126321
Mid-range product       673183
High-range product      417678
Name: count, dtype: int64

### If-Statements with For-Loops

##### To find out how busy each day of the week is?

In [20]:
df_ords_prods_merge['order_days_of_week'].value_counts()

order_days_of_week
0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: count, dtype: int64

In [21]:
# create a list to store the results from the for-loop
result = []

In [22]:
# usage of for-loop
for value in df_ords_prods_merge['order_days_of_week']:
    if value == 0:
        result.append('Busiest Day')
    elif value == 4:
        result.append('Least Busy')
    else:
        result.append('Regularly Busy')

In [23]:
# print the result
result

['Regularly Busy',
 'Regularly Busy',
 'Regularly Busy',
 'Regularly Busy',
 'Regularly Busy',
 'Regularly Busy',
 'Regularly Busy',
 'Regularly Busy',
 'Regularly Busy',
 'Regularly Busy',
 'Regularly Busy',
 'Regularly Busy',
 'Regularly Busy',
 'Regularly Busy',
 'Regularly Busy',
 'Regularly Busy',
 'Least Busy',
 'Least Busy',
 'Least Busy',
 'Least Busy',
 'Least Busy',
 'Least Busy',
 'Least Busy',
 'Least Busy',
 'Least Busy',
 'Least Busy',
 'Least Busy',
 'Least Busy',
 'Least Busy',
 'Regularly Busy',
 'Regularly Busy',
 'Regularly Busy',
 'Regularly Busy',
 'Regularly Busy',
 'Regularly Busy',
 'Regularly Busy',
 'Regularly Busy',
 'Regularly Busy',
 'Regularly Busy',
 'Regularly Busy',
 'Regularly Busy',
 'Regularly Busy',
 'Regularly Busy',
 'Regularly Busy',
 'Regularly Busy',
 'Regularly Busy',
 'Regularly Busy',
 'Regularly Busy',
 'Regularly Busy',
 'Regularly Busy',
 'Least Busy',
 'Least Busy',
 'Least Busy',
 'Least Busy',
 'Least Busy',
 'Least Busy',
 'Least Busy

In [24]:
# create a new column 'Busiest_day' in the df_ords_prods_merge
df_ords_prods_merge['Busiest_day'] = result

In [25]:
df_ords_prods_merge['Busiest_day'].value_counts()

Busiest_day
Regularly Busy    22416875
Busiest Day        6204182
Least Busy         3783802
Name: count, dtype: int64

In [26]:
df_ords_prods_merge.head()

order_id  user_id  order_number  order_days_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     7.0         196                  1          0   
1                     7.0       14084                  2          0   
2                     7.0       12427                  3          0   
3                     7.0       26088                  4          0   
4                     7.0       26405                  5          0   

                              product_name  aisle_id  department_id  prices  \
0                                     Soda        77              7     9.0   
1  Organic Unsweetened Vanilla Almond Milk        91             16    12.5   
2                      Original Beef Jerky        23             19     4.4   
3               Aged White Cheddar Popcorn        23             19     4.7   
4         XL Pick-A-Size Paper Towel Rolls        54             17     1.0   

  _merge    price_range_loc     Busiest_day  
0   both  Mid-range product  Regularly Busy  
1   both  Mid-range product  Regularly Busy  
2   both  Low-range product  Regularly Busy  
3   both  Low-range product  Regularly Busy  
4   both  Low-range product  Regularly Busy

# Task 4.7: Step 2 and 3

In [27]:
# counting the values in the order_days_of_week column
df_ords_prods_merge['order_days_of_week'].value_counts()

order_days_of_week
0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: count, dtype: int64

#### Here the two busiest days are 0 (Saturday) and 1 (Sunday) and the two least busy days are 4 (Wednesday) and 3 (Tuesday)

In [28]:
#Renaming the column Busiest_day to Busiest_days

In [29]:
df_ords_prods_merge.rename(columns = {'Busiest_day' : 'Busiest_days'}, inplace = True)

In [30]:
# create a list to store the values
result1 = []

In [31]:
# usage of for-loop
for value in df_ords_prods_merge['order_days_of_week']:
    if (value == 0) or (value == 1):
        result1.append('Busiest Days')
    elif (value == 4) or (value == 3):
        result1.append('Least Busy Days')
    else:
        result1.append('Regularly Busy')

In [32]:
result1

['Regularly Busy',
 'Regularly Busy',
 'Regularly Busy',
 'Regularly Busy',
 'Regularly Busy',
 'Least Busy Days',
 'Least Busy Days',
 'Least Busy Days',
 'Least Busy Days',
 'Least Busy Days',
 'Least Busy Days',
 'Least Busy Days',
 'Least Busy Days',
 'Least Busy Days',
 'Least Busy Days',
 'Least Busy Days',
 'Least Busy Days',
 'Least Busy Days',
 'Least Busy Days',
 'Least Busy Days',
 'Least Busy Days',
 'Least Busy Days',
 'Least Busy Days',
 'Least Busy Days',
 'Least Busy Days',
 'Least Busy Days',
 'Least Busy Days',
 'Least Busy Days',
 'Least Busy Days',
 'Regularly Busy',
 'Regularly Busy',
 'Regularly Busy',
 'Regularly Busy',
 'Busiest Days',
 'Busiest Days',
 'Busiest Days',
 'Busiest Days',
 'Busiest Days',
 'Busiest Days',
 'Busiest Days',
 'Busiest Days',
 'Busiest Days',
 'Busiest Days',
 'Busiest Days',
 'Busiest Days',
 'Busiest Days',
 'Busiest Days',
 'Busiest Days',
 'Busiest Days',
 'Busiest Days',
 'Least Busy Days',
 'Least Busy Days',
 'Least Busy Days',


In [33]:
df_ords_prods_merge['Busiest_days'] = result1

In [34]:
df_ords_prods_merge['Busiest_days'].value_counts()

Busiest_days
Regularly Busy     12916111
Busiest Days       11864412
Least Busy Days     7624336
Name: count, dtype: int64

In [35]:
df_ords_prods_merge.head()

order_id  user_id  order_number  order_days_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     7.0         196                  1          0   
1                     7.0       14084                  2          0   
2                     7.0       12427                  3          0   
3                     7.0       26088                  4          0   
4                     7.0       26405                  5          0   

                              product_name  aisle_id  department_id  prices  \
0                                     Soda        77              7     9.0   
1  Organic Unsweetened Vanilla Almond Milk        91             16    12.5   
2                      Original Beef Jerky        23             19     4.4   
3               Aged White Cheddar Popcorn        23             19     4.7   
4         XL Pick-A-Size Paper Towel Rolls        54             17     1.0   

  _merge    price_range_loc    Busiest_days  
0   both  Mid-range product  Regularly Busy  
1   both  Mid-range product  Regularly Busy  
2   both  Low-range product  Regularly Busy  
3   both  Low-range product  Regularly Busy  
4   both  Low-range product  Regularly Busy

### Answer: According to the frequency destribution 
##### Regularly Busy:   12916111 
##### Busiest Day:     1186441 
##### LeastBusyDays :7624336

Total count is same as the Total frequency count before categorizing into Busiest Days,Least Busy Days and Regularly Busy6

# Step 4 and 5

In [36]:
# checking the contents in the column order_hour_of_day
df_ords_prods_merge['order_hour_of_day'].value_counts()

order_hour_of_day
10    2761760
11    2736140
14    2689136
15    2662144
13    2660954
12    2618532
16    2535202
9     2454203
17    2087654
8     1718118
18    1636502
19    1258305
20     976156
7      891054
21     795637
22     634225
23     402316
6      290493
0      218769
1      115700
5       87961
2       69375
4       53242
3       51281
Name: count, dtype: int64

## Classifying the hours of the day into 3 categories
#### Most_orders     :- contains top 8 hours having highest value counts (10,11,14,15,13,12,16,9)
#### Average_orders  :- contains middle 8 hours (17,8,18,19,20,7,21,22)
#### Fewest_orders   :- contais the last 8 hours with lowest value counts (23,6,0,1,5,2,4,3)

In [37]:
# use If-Statemenst with for-loop

result2 = []

for value in df_ords_prods_merge["order_hour_of_day"]:
  if value in [10, 11, 14, 15, 13, 12, 16, 9]:
    result2.append("Most orders")
  elif value in [23, 6, 0, 1, 5, 2, 4, 3]:
    result2.append("Fewest orders")
  else:
    result2.append("Average orders")

In [38]:
#printing the result
result2

['Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Avera

In [39]:
#create a new column named busiest_period_of_day and assign the value of result2 to it
df_ords_prods_merge['busiest_period_of_day'] = result2

In [41]:
#Print the frequency for this new column
df_ords_prods_merge['busiest_period_of_day'].value_counts() 

busiest_period_of_day
Most orders       21118071
Average orders     9997651
Fewest orders      1289137
Name: count, dtype: int64

## Step 7 : Export the dataframe as a pickle file

In [42]:
#exporting the dataframe df_ords_prods_merge to a pickle file
df_ords_prods_merge.to_pickle(os.path.join(path, 'Data','Prepared Data', 'orders_products_merged_derived.pkl'))